### PR TITLE
tkt-63801: Check snapshot exists before destroying

### PR DESCRIPTION
**Pull request by sonicaj**
This commit ensures that we check a snapshot exists before we destroy it and log an appropriate message accordingly. It also introduces a snapshot class which we can use to better manage snapshot related operations.
Ticket: #63801

### CODE REVIEW

File: iocage_lib/ioc_json.py [hunk 1]
```diff
@@ -45,7 +45,76 @@
...
+    def normalize_data(self):
+        for line in self.raw_data.splitlines()[1:]:
+            self.snapshots.append(IOCSnapshot(line))
+
+    def __iter__(self):
+        for snap in self.snapshots:
+            # FIXME: I will still hold old snapshot data, let's change this pls
+            yield snap
+
+    def get_snap(self, snap_id):
```
Comment (william-gr) [inline iocage_lib/ioc_json.py:134]: Listing all snapshots to get a single one is a bad idea, specially on system with thousands of them.

File: iocage_lib/ioc_exceptions.py [hunk 2]
```diff
@@ -64,3 +65,25 @@ class JailMissingConfiguration(JailMisconfigured):
 
 class ValidationFailed(ExceptionWithMsg):
     pass
+
+
+@contextmanager
+def ignore_exceptions(
+        *exceptions, action=None, action_args=None,
+        action_kwargs=None, suppress_exception=True
```
Comment (william-gr) [inline iocage_lib/ioc_exceptions.py]: 8 spaces indentation?

File: iocage_lib/ioc_json.py [hunk 3]
```diff
@@ -45,7 +45,76 @@
...
+        if self.raw_data:
+            # Expected format
+            # ['NAME', 'USED', 'AVAIL', 'REFER', 'MOUNTPOINT']
+            self.__dict__.update({
+                k: v for k, v in zip(self.attr_list, self.raw_data.split())
+            })
+
+    def delete(self, recursive=True):
+        with ioc_exceptions.ignore_exceptions(
+                su.CalledProcessError, action=False
```
Comment (william-gr) [inline iocage_lib/ioc_json.py]: 8 spaces indentation?

File: iocage_lib/ioc_json.py [hunk 4]
```diff
@@ -45,7 +45,76 @@
...
+
+class IOCSnapshots:
+    def __init__(self):
+        self.snapshots = []
+        self.normalize_data()
+
+    @property
+    def raw_data(self):
+        with ioc_exceptions.ignore_exceptions(
+                su.CalledProcessError, action=''
```
Comment (william-gr) [inline iocage_lib/ioc_json.py]: 8 spaces indentation?

File: iocage_lib/ioc_exceptions.py [hunk 5]
```diff
@@ -64,3 +65,25 @@ class JailMissingConfiguration(JailMisconfigured):
 
 class ValidationFailed(ExceptionWithMsg):
     pass
+
+
+@contextmanager
+def ignore_exceptions(
+        *exceptions, action=None, action_args=None,
```
Comment (william-gr) [inline iocage_lib/ioc_exceptions.py]: action is a confusing name IMHO. The only use cases you have so far are static return values, and thats what this really is, a `return_value`

If a callable is a edge cage perhaps it makes sense to only use the callable and no args and kwargs. These could be passed using functools.partial if needed.

**Review by skarekrow — APPROVED** [review]

**Pull request merged**